packers/jws.go

```diff
@@ -0,0 +1,384 @@
+		len(didDoc.VerificationMethod)+len(didDoc.Authentication))
+	for i := range didDoc.Authentication {
+		vm, err := resolveAuthToVM(
+			didDoc.Authentication[i],
+			didDoc.VerificationMethod,
+		)
+		if err != nil {
+			continue
+		}
+		vms = append(vms, vm)
+	}
+	vms = append(vms, didDoc.VerificationMethod...)
+
+	if len(vms) == 0 {
+		return verifiable.CommonVerificationMethod{}, errors.New("no verification methods")
+	}
+
+	if kid == "" {
+		return vms[0], nil
+	}
+
+	for i := range vms {
+		if vms[i].ID == kid {
+			return vms[i], nil
+		}
+		if id, ok := vms[i].PublicKeyJwk["kid"]; ok && id == kid {
+			return vms[i], nil
+		}
+	}
+
+	return verifiable.CommonVerificationMethod{}, errors.New("can't find kid")
+}
+
+func resolveAuthToVM(
+	auth verifiable.Authentication,
+	vms []verifiable.CommonVerificationMethod,
+) (verifiable.CommonVerificationMethod, error) {
+	if !auth.IsDID() {
+		return auth.CommonVerificationMethod, nil
+	}
+	// make keys from authenication section more priority
+	for i := range vms {
+		if auth.DID() == vms[i].ID {
+			return vms[i], nil
+		}
+	}
+	return verifiable.CommonVerificationMethod{}, errors.New("not found")
+}
+
+func extractVerifyKey(alg jwa.SignatureAlgorithm, vm verifiable.CommonVerificationMethod) (jws.VerifyOption, error) {
+	supportedAlg, ok := supportedAlgorithms[alg]
+	if !ok {
+		return nil, errors.Errorf("unsupported algorithm: '%s'", alg)
+	}
+	_, ok = supportedAlg[verificationType(vm.Type)]
+	if !ok {
+		return nil, errors.Errorf("unsupported verification type: '%s'", vm.Type)
+	}
+
+	if len(vm.PublicKeyJwk) > 0 {
+		vm.PublicKeyJwk["alg"] = alg
+		bytesJWK, err := json.Marshal(vm.PublicKeyJwk)
+		if err != nil {
+			return nil, errors.Errorf("failed to marshal jwk: %v", err)
+		}
+		jwkKey, err := jwk.ParseKey(bytesJWK)
+		if err != nil {
+			return nil, errors.Errorf("failed to parse jwk: %v", err)
+		}
+		if jwkKey.Algorithm() != BJJAlg {
+			return jws.WithKey(
+				jwkKey.Algorithm(),
+				jwkKey,
+			), nil
+		}
+		bjjKey, err := parseBJJKey(jwkKey)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		return jws.WithKey(
+			jwkKey.Algorithm(),
+			bjjKey,
+		), nil
+	}
+
+	if vm.PublicKeyHex != "" {
+		encodedKey, err := hex.DecodeString(vm.PublicKeyHex)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		return jws.WithKey(
+			jwa.KeyAlgorithmFrom(alg),
+			newecdsa(encodedKey),
+		), nil
+	}
+
+	if vm.PublicKeyBase58 != "" {
+		encodedKey, err := base58.Decode(vm.PublicKeyBase58)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		return jws.WithKey(
+			jwa.KeyAlgorithmFrom(alg),
+			newecdsa(encodedKey),
+		), nil
+	}
+
+	return nil, errors.New("can't find public key")
+}
+
+func newecdsa(encodedKey []byte) ecdsa.PublicKey {
+	return ecdsa.PublicKey{
+		Curve: secp256k1.S256(),
+		X:     new(big.Int).SetBytes(encodedKey[:32]),
+		Y:     new(big.Int).SetBytes(encodedKey[32:]),
+	}
+}
+
+func parseBJJKey(jwkKey jwk.Key) (*bjj.PublicKey, error) {
```

Comment:
not sure that's a right place for these function

Reply:
Any proposal?